Cleanup how text providers are set up and synced

Jacalz opened the pull request:
<!-- If this is your first pull request for Fyne please read the contributor docs at:
https://github.com/fyne-io/fyne/wiki/Contributing.
Be sure that your work is based off `develop` branch. --> 

### Description:
<!-- A summary of the change included and which issue it addresses.
Please include any relevant motivation and background. -->

This cleans up a lot of code for setting up and syncing changes between widgets and their text providers. Text providers are created inline in the struct to not require an extra allocation plus nil checks, segments are synced without needing to allocate garbage in the form of new slices and some other cleanups are included. 

### Checklist:
<!-- Please tick these as appropriate using [x] -->

- [ ] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

dweymouth commented on widget/entry.go:
```diff
@@ -1009,16 +1009,17 @@ func (e *Entry) getRowCol(p fyne.Position) (int, int) {
-	} else if row >= e.textProvider().rows() {
-		row = e.textProvider().rows() - 1
-		col = e.textProvider().rowLength(row)
+	} else if row >= text.rows() {
```
Should we call text.rows() just once?

Jacalz replied:
I checked and doesn't matter. It us just returning the length of a slice and that method will be inlined.

Jacalz commented:
Rebased on latest develop now.

Jacalz commented:
Sorry for the force push. Did a merge first and then another because it still didn't think that the conflicts were resolved. After the second merge, it still didn't believe that I fixed them so I gave up and rebased instead.

Jacalz commented:
I don't understand why the tests are failing like they are. The code worked fine before the merge/rebase.

andydotxyz commented:
> I don't understand why the tests are failing like they are. The code worked fine before the merge/rebase.

No idea. the base branches are all working correctly too.

Jacalz commented:
Will have a look again today or tomorrow. A fresh mind might give some insights

Jacalz commented:
Closing in favour of #4649
